src/lib.rs

```diff
@@ -40,19 +42,410 @@ mod repl_helper;
+        /// * `machine` is not a valid pointer to a life-alive `Machine` instance.
+        /// * `input` is not a null-terminated array.
+        ///
+        /// # Parameters
+        ///
+        /// * `machine`: Raw mutable pointer to a `Machine` instance.
+        /// * `input`:   Raw immutable pointer to a C string containing the Prolog string query.
+        ///
+        /// # Returns
+        ///
+        /// A raw pointer (`*mut QueryState`) to the created `QueryState` or a null pointer in case of a panic.
+        #[no_mangle]
+        pub extern "C" fn scryer_start_new_query_generator(machine: *mut Machine, input: *const c_char) -> *mut QueryState {
+            let result = std::panic::catch_unwind(|| {
+                let c_str: &CStr;
+                unsafe {
+                    assert!(!input.is_null());
+                    c_str = CStr::from_ptr(input);
+                }
+
+                let buf = c_str.to_str().expect("Not a valid UTF-8 string");
+                buf.to_owned()
+            });
+
+            match result {
+                Ok(r_str) => {
+                    let query_state;
+                    unsafe {
+                        query_state = (*machine).start_new_query_generator(r_str);
+                    }
+                    Box::into_raw(Box::new(query_state))
+                }
+                Err(e) => {
+                    eprintln!("Panic: {:?}", e);
+                    std::ptr::null_mut()
+                }
+            }
+        }
+
+        /// This is an FFI function that deallocates a `QueryState` that was previously allocated, then runs the `cleanup_query_generator()` operation on the given `Machine`.
+        /// It returns a pointer to a C string that contains a JSON string indicating the status of the operation ("{"status": "ok"}").
+        ///
+        /// # Safety
+        /// This is an unsafe function because it dereferences raw pointers.
+        /// It is the caller's responsibility to ensure these pointers are valid.
+        /// Moreover, any mutation or deallocation of `machine` and `query_state` after this operation could result in segmentation faults,
+        /// and the C string pointer returned must be properly deallocated using `CString::from_raw()` to avoid a memory leak.  You can use
+        /// the provided `scryer_free_c_string()` to do this.
+        ///
+        /// # Parameters
+        /// * `machine`: a pointer to a `Machine`. This Machine's `cleanup_query_generator()` method will be run.
+        /// * `query_state`: a pointer to a `QueryState`. This will be deallocated during the function.
+        ///
+        /// # Returns
+        ///
+        /// Returns a pointer to a C string ({`status`:"ok"})
+        /// This should be deallocated using CString::from_raw.
+        #[no_mangle]
+        pub extern "C" fn scryer_cleanup_query_generator(machine: *mut Machine, query_state: *mut QueryState) -> *mut c_char {
```

Comment:
While I think that not having a C-style API or supporting residual goals is fine for a MVP, I do value forward compatibility, so at least for the APIs (like `scryer_cleanup_query_generator()` and `scryer_load_module_string()`) that will be shared with the C-style API, using C-style instead of webdev JSON-style errors is a must.

For `scryer_run_query_generator()` (the JSON version) I changed my mind and I think having webdev JSON-style errors is fine, because you will have to use JSON anyway.